### DISCUSSION TITLE
Non-root charon bringup

### DISCUSSION THREAD

**Discussion by shanmukhasreedhar** [strongswan/strongswan#1738] [Q&A]
Hi,
1. We are using strongswan version 5.7.2. For non-root charon bringup, we have tried the steps given in the below link but no luck. We have used charon.user and charon.group in strongswan.conf.
Is there any detailed or other procedure to run charon as non-root?
https://docs.strongswan.org/docs/5.9/install/reducedPrivileges.html

2. After compiling the strongswan with custom options (like ./configure --with-user=vpn --with-group=vpn) to create rpm out of it, do we have any standard procedure? Because I didnt find any rpm spec file in strongswan code base.

**Comment by tobiasbrunner** [comment]
> We are using strongswan version 5.7.2.

Why? That's really old.

> For non-root charon bringup, we have tried the steps given in the below link but no luck.

What does that mean exactly?

> After compiling the strongswan with custom options (like ./configure --with-user=vpn --with-group=vpn)

Setting these options has basically the same effect as changing the settings in strongswan.conf (it's just the default value for these). Either way, it causes the daemon to switch to that user/group when started as root.

Also, if you build from sources, I highly recommend using a more recent version.

> do create rpm out of it, do we have any standard procedure? Because I didnt find any rpm spec file in strongswan code base.

There is none. You could check the Fedora/EPEL package sources (see e.g. [here](https://src.fedoraproject.org/rpms/strongswan)).

**Reply by shanmukhasreedhar** [reply]
Thanks for reply.

 tobiasbrunner>>> What does that mean exactly?

When we brought up Charon binary, with non-root and with particular capabilities as mentioned below we are getting errors and tunnel is not established. 

**Capabilities:**
bash-4.2$ getcap /usr/libexec/strongswan/charon
/usr/libexec/strongswan/charon = **cap_net_bind_service,cap_net_admin,cap_net_raw,cap_audit_write+eip**. 

**Error from strace output against charon binary**

recvfrom(9, {{len=40, type=NLMSG_ERROR, flags=0, seq=201, pid=19382}, {**error=-EPERM**, msg={{len=20, type=XFRM_MSG_GETSPDINFO, flags=NLM_F_REQUEST, seq=201, pid=19382}, "\x00\x00\x00\x00"}}}, 4096, MSG_TRUNC, NULL, NULL) = 40
recvfrom(9, {{len=40, type=NLMSG_ERROR, flags=0, seq=202, pid=19382}, {**error=-EPERM**, msg={{len=20, type=XFRM_MSG_GETSPDINFO, flags=NLM_F_REQUEST, seq=202, pid=19382}, "\x00\x00\x00\x00"}}}, 4096, MSG_TRUNC, NULL, NULL) = 40
bind(10, {sa_family=AF_NETLINK, nl_pid=0, nl_groups=0x0000c3}, 12) = -1 EPERM (Operation not permitted)
recvfrom(0, {{len=56, type=NLMSG_ERROR, flags=0, seq=203, pid=19382}, {**error=-EPERM**, msg={{len=36, type=RTM_NEWRULE, flags=NLM_F_REQUEST|NLM_F_ACK|NLM_F_EXCL|NLM_F_CREATE, seq=203, pid=19382}, {family=AF_INET, dst_len=0, src_len=0, tos=0, table=0xdc, action=FR_ACT_TO_TBL, flags=0}, {{nla_len=8, nla_type=FRA_PRIORITY}, 220}}}}, 4096, MSG_TRUNC, NULL, NULL) = 56
recvfrom(0, {{len=56, type=NLMSG_ERROR, flags=0, seq=204, pid=19382}, {**error=-EPERM**, msg={{len=36, type=RTM_NEWRULE, flags=NLM_F_REQUEST|NLM_F_ACK|NLM_F_EXCL|NLM_F_CREATE, seq=204, pid=19382}, {family=AF_INET6, dst_len=0, src_len=0, tos=0, table=0xdc, action=FR_ACT_TO_TBL, flags=0}, {{nla_len=8, nla_type=FRA_PRIORITY}, 220}}}}, 4096, MSG_TRUNC, NULL, NULL) = 56
recvfrom(0, {{len=56, type=NLMSG_ERROR, flags=0, seq=205, pid=19382}, {**error=-EPERM**, msg={{len=36, type=RTM_DELRULE, flags=NLM_F_REQUEST|NLM_F_ACK, seq=205, pid=19382}, {family=AF_INET, dst_len=0, src_len=0, tos=0, table=0xdc, action=FR_ACT_TO_TBL, flags=0}, {{nla_len=8, nla_type=FRA_PRIORITY}, 220}}}}, 4096, MSG_TRUNC, NULL, NULL) = 56
recvfrom(0, {{len=56, type=NLMSG_ERROR, flags=0, seq=206, pid=19382}, {**error=-EPERM**, msg={{len=36, type=RTM_DELRULE, flags=NLM_F_REQUEST|NLM_F_ACK, seq=206, pid=19382}, {family=AF_INET6, dst_len=0, src_len=0, tos=0, table=0xdc, action=FR_ACT_TO_TBL, flags=0}, {{nla_len=8, nla_type=FRA_PRIORITY}, 220}}}}, 4096, MSG_TRUNC, NULL, NULL) = 56

**Reply by tobiasbrunner** [reply]
Note that using `strace` will probably only work with the `-u <username>` option and if started as root (e.g. via `sudo`), see NOTES in the capabilities(7) man page.

Also, have you compiled with `--with-capabilities=<native|libcap>`? Otherwise, the daemon won't be able to verify that it has the required capabilities (without capabilities support it simply checks the uid).
